# 유튜브 인기동영상 데이터
데이터 출처 :https://www.kaggle.com/rsrishav/youtube-trending-video-dataset?select=KR_youtube_trending_data.csv  
데이터 설명 : 유튜브 데일리 인기동영상 (한국)  
dataurl : https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/youtube.csv

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/youtube.csv")
df.head()

Unnamed: 0                                              title  \
0           0                                          [신병] 물자창고   
1           1   RAIN(비) - 나로 바꾸자 Switch to me (duet with JYP) MV   
2           2       2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.   
3           3                                       고기남자의 칠면조 파티   
4           4  골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...   

              channelTitle  categoryId  view_count  likes  dislikes  \
0                      장삐쭈          23     1893473  38249       730   
1  RAIN's Official Channel          10     2600864      0         0   
2              서울시 · Seoul          29      347049   3564       120   
3             고기남자 MeatMan          26      528458  15372       280   
4                    스브스밥집          24      494904   3918       111   

   comment_count                 channelId trending_date2  
0           8595  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
1          20129  UCxXgIeE5hxWxHG6dz9Scg2w     2021-01-01  
2            178  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3           3470  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4           3142  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01

##### 1. 인기동영상 제작횟수가 많은 채널 상위 10개명을 출력하라 (날짜기준, 중복포함)

In [3]:
df["channelId"].isin(df["channelId"].value_counts().head(10).index)

0         True
1        False
2        False
3        False
4        False
         ...  
60394    False
60395    False
60396    False
60397    False
60398    False
Name: channelId, Length: 60399, dtype: bool

In [4]:
df[df["channelId"].isin(df["channelId"].value_counts().head(10).index)]["channelTitle"].unique()

array(['장삐쭈', '총몇명', '파뿌리', '짤툰', '런닝맨 - 스브스 공식 채널',
       '엠뚜루마뚜루 : MBC 공식 종합 채널', 'SPOTV', '채널 십오야', '이과장', 'BANGTANTV'],
      dtype=object)

##### 2. 논란으로 인기동영상이 된 케이스를 확인하고 싶다. dislikes수가 like 수보다 높은 동영상을 제작한 채널을 모두 출력하라

In [5]:
df[df["dislikes"] > df["likes"]].channelTitle.unique()

array(['핫도그TV', 'ASMR 애정TV', '하얀트리HayanTree', '양팡 YangPang',
       '철구형 (CHULTUBE)', '왜냐맨하우스', '(MUTUBE)와꾸대장봉준', '오메킴TV', '육지담',
       'MapleStory_KR', 'ROAD FIGHTING CHAMPIONSHIP', '사나이 김기훈',
       '나혼자산다 STUDIO', 'Gen.G esports'], dtype=object)

##### 3. 채널명을 바꾼 케이스가 있는지 확인하고 싶다. channelId의 경우 고유값이므로 이를 통해 채널명을 한번이라도 바꾼 채널의 갯수를 구하여라

In [6]:
df[["channelTitle", "channelId"]]

channelTitle                 channelId
0                          장삐쭈  UChbE5OZQ6dRHECsX0tEPEZQ
1      RAIN's Official Channel  UCxXgIeE5hxWxHG6dz9Scg2w
2                  서울시 · Seoul  UCZUPZW5idAxYp-Asj__lVAA
3                 고기남자 MeatMan  UCT3CumbFIJiW33uq0UI3zlg
4                        스브스밥집  UCdWgRSfttvDucq4ApcCg5Mw
...                        ...                       ...
60394                       공룡  UCWf4QKgnTyvmehcRqw-JldQ
60395                  놀면 뭐하니?  UCx6jsZ02B4K3SECUrkgPyzg
60396        김종국 GYM JONG KOOK  UCoe-0EVDJnjlSoPK8ygcGwQ
60397                 스브스 예능맛집  UCnx4Fi4cmkLGDiFfZ311wWw
60398              HYBE LABELS  UC3IZKseVpdzPSBaWxBxundA

[60399 rows x 2 columns]

In [7]:
df[["channelTitle", "channelId"]].drop_duplicates()["channelId"].value_counts()

channelId
UC98TOxKQk4aLcx0EjIK0LkQ    3
UCtQYhzJWFai1y30UIA3uwYQ    3
UCiEEF51uRAeZeCo8CJFhGWw    2
UCytBYTjlE45XxrJ7SxO2igg    2
UC4LW3sOslC14Vn5bcoNUBkg    2
                           ..
UCeAVJesn059sG6t-1XCgfJQ    1
UCureTzUsWPyrhfZ38x31WXQ    1
UCFXigv26l81ZBaKz6RakL0A    1
UC7lP46vHWIdZbT0X7Ud0AAA    1
UCOH52Yqq4-rdLvpt2Unsqsw    1
Name: count, Length: 1770, dtype: int64

In [8]:
ans = df[["channelTitle", "channelId"]].drop_duplicates()["channelId"].value_counts()
len(ans[ans > 1])

71

##### 4. 일요일에 인기있었던 영상들중 가장많은 영상 종류(categoryId)는 무엇인가?

In [9]:
import datetime as dt

In [10]:
df["trending_date2"] = pd.to_datetime(df["trending_date2"])

In [11]:
df[df["trending_date2"].dt.day_name() == "Sunday"]["categoryId"].value_counts()

categoryId
24    3096
22    1337
10     830
17     636
23     556
25     437
26     363
20     289
1      274
15     217
27     199
28     167
2       99
19      91
29       9
Name: count, dtype: int64

In [12]:
df[df["trending_date2"].dt.day_name() == "Sunday"]["categoryId"].value_counts().index[0]

np.int64(24)

##### 5. 각 요일별 인기 영상들의 categoryId는 각각 몇개 씩인지 하나의 데이터 프레임으로 표현하라

In [13]:
ans = df.groupby([df["trending_date2"].dt.day_name(), "categoryId"], as_index = False).size()
ans.pivot(index = "categoryId", columns = "trending_date2")

size                                                  
trending_date2 Friday Monday Saturday Sunday Thursday Tuesday Wednesday
categoryId                                                             
1                 243    263      255    274      246     257       234
2                 120    105      119     99      128     119       129
10                833    837      776    830      890     894       917
15                187    215      198    217      207     208       207
17                633    668      592    636      682     708       706
19                 90     92       87     91       92      89        85
20                283    298      296    289      282     285       291
22               1288   1373     1289   1337     1341    1375      1333
23                568    594      570    556      560     569       566
24               2976   3148     3066   3096     2954    3084      3090
25                444    453      422    437      470     452       468
26                369    378      364    363      375     394       385
27                183    205      183    199      194     194       212
28                171    160      173    167      166     161       165
29                 12     10       10      9       13      11        12

##### 6. 댓글의 수로 (comment_count) 영상 반응에 대한 판단을 할 수 있다. viewcount대비 댓글수가 가장 높은 영상을 확인하라 (view_count값이 0인 경우는 제외한다)

In [14]:
target = df[df["view_count"] != 0]

In [15]:
target["ratio"] = target["comment_count"] / target["view_count"]
ans = target.sort_values("ratio", ascending = False).dropna()
ans.iloc[0].title

C:\Users\cho03\AppData\Local\Temp\ipykernel_31996\589280512.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  target["ratio"] = target["comment_count"] / target["view_count"]


'60분 동안 댓글이 달리지 않으면, 영상이 삭제됩니다. (챌린지)'

##### 7. 댓글의 수로 (comment_count) 영상 반응에 대한 판단을 할 수 있다.viewcount대비 댓글수가 가장 낮은 영상을 확인하라 (view_counts, ratio값이 0인경우는 제외한다.)

In [16]:
target2 = target[target["ratio"] != 0]
ans = target2.sort_values("ratio")
ans.iloc[0].title

'Join the BTS #PermissiontoDance Challenge only on YouTube #Shorts'

##### 8. like 대비 dislike의 수가 가장 적은 영상은 무엇인가? (like, dislike 값이 0인경우는 제외한다)

In [17]:
target3 = df[(df.likes != 0) & (df.dislikes != 0)]
df.iloc[(target3["dislikes"] / target3["likes"]).sort_values().index[0]].title

'[줌터뷰] *최초공개* 사부작즈🐰🐶의 비공식 이름은 아이라인즈? 꿀조합 티키타카 가득한 NCT 127 도영&정우의 줌터뷰'

##### 9. 가장많은 트렌드 영상을 제작한 채널의 이름은 무엇인가? (날짜기준, 중복포함)

In [18]:
df["channelId"].value_counts().index[0]

'UCszFjh7CEfwDb7UUGb4RzCQ'

In [19]:
df.loc[df["channelId"] == df["channelId"].value_counts().index[0]].channelTitle.unique()[0]


'짤툰'

##### 10. 20회(20일)이상 인기동영상 리스트에 포함된 동영상의 숫자는?

In [20]:
(df[["title", "channelId"]].value_counts() >= 20).sum()

np.int64(40)

# ❌ 유튜브 공범컨텐츠 동영상 데이터
데이터 출처 :https://www.kaggle.com/kukuroo3/youtube-episodic-contents-kr(참고, 데이터 수정)  
데이터 설명 : 유튜브 “공범” 컨텐츠 동영상 정보 ( 10분 간격 수집)  
dataurl1 (비디오 정보) = https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/videoInfo.csv  
dataurl2 (참가자 채널 정보)= https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/channelInfo.csv  

In [21]:
video = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/videoInfo.csv")
channel = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/channelInfo.csv")
display(video.head())
display(channel.head())

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
0  c5JQp6xafqc  1667010    30474         706            0   6587   
1  c5JQp6xafqc  1669089    30495         707            0   6589   
2  c5JQp6xafqc  1674759    30522         711            0   6596   
3  c5JQp6xafqc  1677026    30555         712            0   6604   
4  c5JQp6xafqc  1681824    30585         713            0   6600   

                    ct videoname  
0  2021-10-10 15:20:03    공범 EP1  
1  2021-10-10 15:30:03    공범 EP1  
2  2021-10-10 15:40:02    공범 EP1  
3  2021-10-10 15:50:03    공범 EP1  
4  2021-10-10 16:00:03    공범 EP1

channelid   subcnt    viewcnt  videocnt  \
0  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  410238653       736   
1  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       736   
2  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       735   
3  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       737   
4  UCkQCwnkQfgSuPTTnw_Y7v7w  1320000  412531322       737   

                    ct channelname  
0  2021-09-30 03:01:03         꽈뚜룹  
1  2021-09-30 09:01:03         꽈뚜룹  
2  2021-09-30 15:01:03         꽈뚜룹  
3  2021-09-30 21:01:03         꽈뚜룹  
4  2021-10-01 03:01:04         꽈뚜룹

##### 11. 각 데이터의 ‘ct’컬럼을 시간으로 인식할수 있게 datatype을 변경하고 video 데이터의 videoname의 각 value 마다 몇개의 데이터씩 가지고 있는지 확인하라

In [22]:
video["ct"] = pd.to_datetime(video["ct"])
channel["ct"] = pd.to_datetime(channel["ct"])

video["videoname"].value_counts()

videoname
공범 EP1    3492
공범 EP2    3204
공범 EP3    2568
공범 EP4    2280
공범 EP5    1562
공범 EP6    1274
공범 EP7     555
공범 EP8     266
Name: count, dtype: int64

##### 12. 수집된 각 video의 가장 최신화 된 날짜의 viewcount값을 출력하라

In [23]:
video.sort_values(["videoname","ct"]).drop_duplicates("videoname", keep = "last")

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
3491   c5JQp6xafqc  3180532    41043        1113            0   7674   
6695   23QQ1ru9YQg  2199328    25566         763            0   9171   
9263   GygoAk1hDU0  1671294    22719         508            0  12499   
11543  bK0_3Vbfvs8  1818493    22703         408            0  16949   
13105  hv7FBjskAu0  1503435    19926         421            0  12434   
14934  AxGKHmXyuAE  1750222    24494         389            0  13025   
13660  QUjQbgj5IXM  1630200    29335         400            0  18247   
15200  yZt-h-KcmUE  1289088    25616        3621            0  31663   

                       ct videoname  
3491  2021-11-01 15:30:03    공범 EP1  
6695  2021-11-01 15:30:03    공범 EP2  
9263  2021-11-01 15:30:03    공범 EP3  
11543 2021-11-01 15:30:03    공범 EP4  
13105 2021-11-01 15:30:04    공범 EP5  
14934 2021-11-01 15:30:04    공범 EP6  
13660 2021-11-01 15:30:05    공범 EP7  
15200 2021-11-01 15:30:05    공범 EP8

##### 13. Channel 데이터중 2021-10-03일 이후 각 채널의 처음 기록 됐던 구독자 수(subcnt)를 출력하라

In [24]:
ans = channel[channel["ct"] >= pd.to_datetime("2021-10-03")].sort_values(["ct", "channelname"]).drop_duplicates().reset_index(drop = True)
ans[["channelname", "subcnt"]]

channelname   subcnt
0            논리왕 전기   922000
1             야전삽짱재   257000
2        츄정ChuJeong   322000
3        김농밀의 농밀한 삶     7520
4       릴펄 Lilpearl    10100
...             ...      ...
1283  곽토리 kwak tori   469000
1284     김농밀의 농밀한 삶     9060
1285    릴펄 Lilpearl    21100
1286  Balming Tiger    56800
1287            꽈뚜룹  1400000

[1288 rows x 2 columns]

##### 14. 각채널의 2021-10-03 03:00:00 ~ 2021-11-01 15:00:00 까지 구독자수 (subcnt) 의 증가량을 구하여라

In [25]:
start = pd.Timestamp("2021-10-03 03:00")
end = pd.Timestamp("2021-11-01 15:00")
target = channel[(channel["ct"] >= start) & (channel["ct"] <= end)]
ans = target.sort_values("ct").groupby("channelname").agg(start = ("subcnt", "first"), end = ("subcnt", "last"))
ans["increase"] = ans["end"] - ans["start"]
ans

start      end  increase
channelname                                
Balming Tiger      54300    56800      2500
곽토리 kwak tori     471000   469000     -2000
김농밀의 농밀한 삶          7520     9060      1540
꽈뚜룹              1330000  1400000     70000
논리왕 전기            922000   911000    -11000
릴펄 Lilpearl        10100    21100     11000
야전삽짱재             257000   268000     11000
와글와글 WagleWagle    55000    55000         0
조나단               215000   227000     12000
츄정ChuJeong        322000   323000      1000
형사!탐정되다            14900    25200     10300

In [26]:
ans.reset_index()[["channelname", "increase"]]

channelname  increase
0     Balming Tiger      2500
1     곽토리 kwak tori     -2000
2        김농밀의 농밀한 삶      1540
3               꽈뚜룹     70000
4            논리왕 전기    -11000
5       릴펄 Lilpearl     11000
6             야전삽짱재     11000
7   와글와글 WagleWagle         0
8               조나단     12000
9        츄정ChuJeong      1000
10          형사!탐정되다     10300

##### 15. 각 비디오는 10분 간격으로 구독자수, 좋아요, 싫어요수, 댓글수가 수집된것으로 알려졌다. 공범 EP1의 비디오정보 데이터중 수집간격이 5분 이하, 20분이상인 데이터 구간( 해당 시점 전,후) 의 시각을 모두 출력하라

In [27]:
target = video[video["videoname"].str.contains("공범 EP1")].sort_values("ct").reset_index(drop = True)
target[(target["ct"].diff() >= dt.timedelta(minutes = 20)) | (target["ct"].diff() <= dt.timedelta(minutes = 5))]

target[target.index.isin([720,721,722,723,1635,1636,1637])]

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
720   c5JQp6xafqc  2153676    34104         830            0   7120   
721   c5JQp6xafqc  2228250    34559         849            0   7191   
722   c5JQp6xafqc  2228250    34559         849            0   7191   
723   c5JQp6xafqc  2228669    34566         849            0   7194   
1635  c5JQp6xafqc  2706691    37893         987            0   7410   
1636  c5JQp6xafqc  2707933    37901         988            0   7410   
1637  c5JQp6xafqc  2708448    37903         988            0   7410   

                      ct videoname  
720  2021-10-12 19:20:03    공범 EP1  
721  2021-10-13 09:41:37    공범 EP1  
722  2021-10-13 09:41:37    공범 EP1  
723  2021-10-13 09:50:03    공범 EP1  
1635 2021-10-19 17:50:02    공범 EP1  
1636 2021-10-19 18:20:03    공범 EP1  
1637 2021-10-19 18:30:03    공범 EP1

##### 16. 각 에피소드의 시작날짜(년-월-일)를 에피소드 이름과 묶어 데이터 프레임으로 만들고 출력하라

In [28]:
start = video.sort_values(["videoname", "ct"]).drop_duplicates("videoname")[["ct", "videoname"]]
start["date"] = start["ct"].dt.date
start[["date", "videoname"]]

date videoname
10     2021-10-07    공범 EP1
3496   2021-10-09    공범 EP2
6696   2021-10-14    공범 EP3
9264   2021-10-16    공범 EP4
11544  2021-10-21    공범 EP5
13661  2021-10-23    공범 EP6
13106  2021-10-28    공범 EP7
14935  2021-10-30    공범 EP8

##### 17. “공범” 컨텐츠의 경우 19:00시에 공개 되는것으로 알려져있다. 공개된 날의 21시의 viewcnt, ct, videoname 으로 구성된 데이터 프레임을 viewcnt를 내림차순으로 정렬하여 출력하라

In [29]:
ans = video[video["ct"].dt.hour == 21].sort_values(["videoname","ct"]).drop_duplicates("videoname").sort_values("viewcnt",ascending=False)[["videoname","viewcnt","ct"]].reset_index(drop=True)
ans

videoname  viewcnt                  ct
0    공범 EP8   264029 2021-10-30 21:00:08
1    공범 EP7   252032 2021-10-28 21:00:04
2    공범 EP4   217674 2021-10-16 21:00:04
3    공범 EP6   213899 2021-10-23 21:00:06
4    공범 EP5   201179 2021-10-21 21:00:04
5    공범 EP2   148144 2021-10-09 21:00:03
6    공범 EP3   147183 2021-10-14 21:00:04
7    공범 EP1   117340 2021-10-07 21:00:03

##### 18. video 정보의 가장 최근 데이터들에서 각 에피소드의 싫어요/좋아요 비율을 ratio 컬럼으로 만들고 videoname, ratio로 구성된 데이터 프레임을 ratio를 오름차순으로 정렬하라

In [30]:
target = video.sort_values("ct").drop_duplicates("videoname",keep="last")
target["ratio"] = target["dislikecnt"] / target["likecnt"]
target.sort_values("ratio")[["videoname", "ratio"]].reset_index(drop = True)

videoname     ratio
0    공범 EP7  0.013636
1    공범 EP6  0.015881
2    공범 EP4  0.017971
3    공범 EP5  0.021128
4    공범 EP3  0.022360
5    공범 EP1  0.027118
6    공범 EP2  0.029844
7    공범 EP8  0.141357

##### 19. 2021-11-01 00:00:00 ~ 15:00:00까지 각 에피소드별 viewcnt의 증가량을 데이터 프레임으로 만드시오

In [31]:
start = pd.to_datetime("2021-11-01 00:00")
end = pd.to_datetime("2021-11-01 15:00")
target = video[(video["ct"] >= start) & (video["ct"] <= end)].sort_values("ct").groupby("videoname").agg(start = ("viewcnt","first"), end = ("viewcnt", "last"))
target["increase"] = target["end"] - target["start"]
target.reset_index()[["videoname", "increase"]]

videoname  increase
0    공범 EP1     13298
1    공범 EP2     10300
2    공범 EP3      9927
3    공범 EP4      9824
4    공범 EP5     10824
5    공범 EP6     14141
6    공범 EP7     26949
7    공범 EP8     89147

##### 20. video 데이터 중에서 중복되는 데이터가 존재한다. 중복되는 각 데이터의 시간대와 videoname 을 구하여라

In [32]:
set(video.index) -  set(video.drop_duplicates().index)

{722, 3927}

In [33]:
answer  = video[video.index.isin(set(video.index) -  set(video.drop_duplicates().index))]
answer[["videoname","ct"]]

videoname                  ct
722     공범 EP1 2021-10-13 09:41:37
3927    공범 EP2 2021-10-13 09:41:37

# ❌ 월드컵 출전선수 골기록 데이터
데이터 출처 :https://www.kaggle.com/darinhawley/fifa-world-cup-goalscorers-19302018(참고, 데이터 수정)  
데이터 설명 : 1930 ~2018년도 월드컵 출전선수 골기록  
data url = https://raw.githubusercontent.com/Datamanim/datarepo/main/worldcup/worldcupgoals.csv

In [34]:
df= pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/worldcup/worldcupgoals.csv")
df.head()

Player  Goals                Years  Country
0  Miroslav Klose     16  2002-2006-2010-2014  Germany
1         Ronaldo     15       1998-2002-2006   Brazil
2     Gerd Muller     14            1970-1974  Germany
3   Just Fontaine     13                 1958   France
4            Pele     12  1958-1962-1966-1970   Brazil

##### 21. 주어진 전체 기간의 각 나라별 골득점수 상위 5개 국가와 그 득점수를 데이터프레임형태로 출력하라

In [35]:
df.groupby("Country").sum().sort_values("Goals", ascending = False).head(5)

Player  Goals  \
Country                                                               
Brazil     RonaldoPeleAdemirJairzinhoVavaLeonidasRivaldoC...    228   
Germany    Miroslav KloseGerd MullerJurgen KlinsmannThoma...    226   
Argentina  Gabriel BatistutaDiego MaradonaGuillermo Stabi...    135   
Italy      Roberto BaggioPaolo RossiChristian VieriSalvat...    123   
France     Just FontaineThierry HenryMichel PlatiniZinedi...    115   

                                                       Years  
Country                                                       
Brazil     1998-2002-20061958-1962-1966-197019501966-1970...  
Germany    2002-2006-2010-20141970-19741990-1994-19982010...  
Argentina  1994-1998-20021982-1986-1990-199419301974-1978...  
Italy      1990-1994-19981978-19821998-2002199019381982-1...  
France     19581998-2002-2006-20101978-1982-19861998-2002...

##### 22. 주어진 전체기간동안 골득점을 한 선수가 가장 많은 나라 상위 5개 국가와 그 선수 숫자를 데이터 프레임 형식으로 출력하라

In [36]:
df.groupby("Country").size().sort_values(ascending = False).head(5)

Country
Brazil       81
Germany      78
Italy        60
Argentina    59
France       58
dtype: int64

##### 23. Years 컬럼은 년도 -년도 형식으로 구성되어있고, 각 년도는 4자리 숫자이다. 년도 표기가 4자리 숫자로 안된 케이스가 존재한다. 해당 건은 몇건인지 출력하라

In [37]:
df["Year_list"] = df["Years"].str.split("-")

def check_year(year):
    for value in year:
        if len(str(value)) != 4: return False

    return True

df["check"] = df["Year_list"].apply(check_year)
len(df[df["check"] == False])

45

##### 24. Q3에서 발생한 예외 케이스를 제외한 데이터프레임을 df2라고 정의하고 데이터의 행의 숫자를 출력하라 (아래 문제부터는 df2로 풀이하겠습니다)

In [38]:
df2 = df[df["check"] == True].reset_index(drop = True)
df2.shape[0]

1250

##### 25. 월드컵 출전횟수를 나타내는 ‘LenCup’ 컬럼을 추가하고 4회 출전한 선수의 숫자를 구하여라

In [39]:
df2["LenCup"] = df2["Year_list"].str.len()
df2["LenCup"].value_counts()[4] # 값이 4인 행이 몇 개인지 인덱싱

np.int64(16)

##### 26. Yugoslavia 국가의 월드컵 출전횟수가 2회인 선수들의 숫자를 구하여라

In [40]:
len(df2[(df2["LenCup"] == 2) & (df2["Country"] == "Yugoslavia")])

7

##### 27. 2002년도에 출전한 전체 선수는 몇 명인가?

In [41]:
len(df2[df2["Years"].str.contains("2002")])

156

##### 28. 이름에 ‘carlos’ 단어가 들어가는 선수의 숫자는 몇 명인가? (대, 소문자 구분 x)

In [42]:
len(df2[df2["Player"].str.lower().str.contains("carlos")])

13

##### 29. 월드컵 출전 횟수가 1회뿐인 선수들 중에서 가장 많은 득점을 올렸던 선수는 누구인가?

In [43]:
df2[df2["LenCup"] == 1].sort_values("Goals", ascending = False)["Player"].values[0]

'Just Fontaine'

##### 30. 월드컵 출전횟수가 1회 뿐인 선수들이 가장 많은 국가는 어디인가?

In [44]:
df2[df2["LenCup"] == 1].groupby("Country").size().sort_values(ascending = False).index[0]

'Brazil'

In [45]:
df2[df2["LenCup"] == 1]["Country"].value_counts().index[0]

'Brazil'

# 서울시 따릉이 이용정보 데이터
데이터 출처 :https://www.data.go.kr/data/15051872/fileData.do(참고, 데이터 수정)  
데이터 설명 : 서울특별시_공공자전거 시간대별 이용정보  
data url = https://raw.githubusercontent.com/Datamanim/datarepo/main/bicycle/seoul_bi.csv

In [46]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/bicycle/seoul_bi.csv")
df.head()

대여일자  대여시간  대여소번호 대여구분코드 성별 연령대코드  이용건수    운동량   탄소량     이동거리  사용시간
0  2021-06-01     0   3541    정기권  F  ~10대     1   0.00  0.00     0.00     8
1  2021-06-01     0    765    정기권  F  ~10대     1  27.21  0.35  1526.81    19
2  2021-06-01     0   2637    정기권  F  ~10대     1  41.40  0.37  1608.56    18
3  2021-06-01     0   2919    정기권  F  ~10대     1   0.00  0.00     0.00    75
4  2021-06-01     0    549    정기권  F  ~10대     1  13.04  0.17   731.55     6

##### 31. 대여일자별 데이터의 수를 데이터프레임으로 출력하고, 가장 많은 데이터가 있는 날짜를 출력하라

In [47]:
res = df["대여일자"].value_counts().sort_index().to_frame()
ans = res.sort_values("count", ascending = False).index[0]

display(res)
print(ans)

count
대여일자             
2021-06-01  48081
2021-06-02  56100
2021-06-03  18985
2021-06-04  56788
2021-06-05  52730
2021-06-06  47787
2021-06-07  52623
2021-06-08  55623
2021-06-09  54507
2021-06-10  36992
2021-06-11  52089
2021-06-12  48569
2021-06-13  43131
2021-06-14  54945

2021-06-04


##### 32. 각 일자의 요일을 표기하고 (‘Monday’ ~’Sunday’) ‘day_name’컬럼을 추가하고 이를 이용하여 각 요일별 이용 횟수의 총합을 데이터 프레임으로 출력하라

In [48]:
df["대여일자"] = pd.to_datetime(df["대여일자"])
df["day_name"]  = df["대여일자"].dt.day_name()
df["day_name"].value_counts().to_frame()

count
day_name         
Wednesday  110607
Friday     108877
Monday     107568
Tuesday    103704
Saturday   101299
Sunday      90918
Thursday    55977

##### 33. 각 요일별 가장 많이 이용한 대여소의 이용횟수와 대여소 번호를 데이터 프레임으로 출력하라

In [49]:
res = df.groupby(["day_name", "대여소번호"]).size().to_frame("size").sort_values(["day_name", "size"], ascending = False).reset_index()
res.drop_duplicates("day_name", keep = "first").reset_index(drop = True)

day_name  대여소번호  size
0  Wednesday    502   282
1    Tuesday    502   267
2   Thursday   2715   137
3     Sunday    502   372
4   Saturday    502   378
5     Monday    502   242
6     Friday    502   277

##### 34. 나이대별 대여구분 코드의 (일일권/전체횟수) 비율을 구한 후 가장 높은 비율을 가지는 나이대를 확인하라. 일일권의 경우 일일권 과 일일권(비회원)을 모두 포함하라

In [50]:
daily = df[df["대여구분코드"].isin(["일일권","일일권(비회원)"])]["연령대코드"].value_counts().sort_index()
total = df["연령대코드"].value_counts().sort_index()

(daily / total).sort_values(ascending = False).index[0]

'~10대'

##### 35. 연령대별 평균 이동거리를 구하여라

In [51]:
df[["연령대코드","이동거리"]].groupby(["연령대코드"]).mean()

이동거리
연령대코드             
20대    3211.890552
30대    3341.443859
40대    3514.857416
50대    3593.668100
60대    3538.145737
70대~   3085.039641
~10대   2634.426279

##### 36. 연령대 코드가 20대인 데이터를 추출하고,이동거리값이 추출한 데이터의 이동거리값의 평균 이상인 데이터를 추출한다.최종 추출된 데이터를 대여일자, 대여소 번호 순서로 내림차순 정렬 후 1행부터 200행까지의 탄소량의 평균을 소숫점 3째 자리까지 구하여라

In [52]:
df2 = df[df["연령대코드"] == "20대"]
res = df2[df2["이동거리"] >= df2["이동거리"].mean()]
res["탄소량"] = res["탄소량"].astype("float")
res = res.sort_values(["대여일자", "대여소번호"], ascending = False).iloc[0:200]["탄소량"]
round(res.sum() / len(res), 3)

C:\Users\cho03\AppData\Local\Temp\ipykernel_31996\379943478.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res["탄소량"] = res["탄소량"].astype("float")


np.float64(1.613)

##### 37. 6월 7일 ~10대의 “이용건수”의 중앙값은?

In [53]:
df["대여일자"] = pd.to_datetime(df["대여일자"])
df[(df["연령대코드"] == "~10대") & (df["대여일자"] == pd.to_datetime("2021-06-07"))]["이용건수"].median()

np.float64(1.0)

##### 38. 평일 (월~금) 출근 시간대(오전 6,7,8시)의 대여소별 이용 횟수를 구해서 데이터 프레임 형태로 표현한 후 각 대여시간별 이용 횟수의 상위 3개 대여소와 이용횟수를 출력하라

In [54]:
df2 = df[(df["day_name"].isin(["Tuesday", "Wednesday", "Thursday", "Friday","Monday"])) & (df["대여시간"].isin([6, 7, 8]))].reset_index()
res = df2.groupby(["대여시간", "대여소번호"]).size().to_frame("이용 횟수")
res.sort_values(["대여시간", "이용 횟수"], ascending = False).groupby("대여시간").head(3)

이용 횟수
대여시간 대여소번호       
8    2701     119
     646      115
     1152      92
7    259      104
     230       77
     726       77
6    2744      45
     1125      40
     1028      36

##### 39. 이동거리의 평균 이상의 이동거리 값을 가지는 데이터를 추출하여 추출데이터의 이동거리의 표본표준편차 값을 구하여라

In [55]:
res = df[df["이동거리"] >= df["이동거리"].mean()]
res["이동거리"].std()

np.float64(5092.1397075053565)

##### 40. 남성(‘M’ or ‘m’)과 여성(‘F’ or ‘f’)의 이동거리값의 평균값을 구하여라

In [56]:
df['sex'] = df['성별'].map(lambda x: '남' if x in ['M','m'] else '여')
df[["sex", "이동거리"]].groupby("sex").mean()

이동거리
sex             
남    3209.110871
여    3468.575025

# 전세계 행복도 지표 데이터
데이터 출처 :https://www.kaggle.com/unsdsn/world-happiness(참고, 데이터 수정)  
데이터 설명 : 전세계 행복도 지표 조사  
data url = https://raw.githubusercontent.com/Datamanim/datarepo/main/happy2/happiness.csv

In [57]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/happy2/happiness.csv", encoding = "utf-8")
df

행복랭킹                       나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도  \
0       1                   Finland  7.769  1.340  1.587  0.986    0.596   
1       2                   Denmark  7.600  1.383  1.573  0.996    0.592   
2       3                    Norway  7.554  1.488  1.582  1.028    0.603   
3       4                   Iceland  7.494  1.380  1.624  1.026    0.591   
4       5               Netherlands  7.488  1.396  1.522  0.999    0.557   
..    ...                       ...    ...    ...    ...    ...      ...   
307   152                     Yemen  3.355  0.442  1.073  0.343    0.244   
308   153                  Tanzania  3.303  0.455  0.991  0.381    0.481   
309   154               South Sudan  3.254  0.337  0.608  0.177    0.112   
310   155  Central African Republic  3.083  0.024  0.000  0.010    0.305   
311   156                   Burundi  2.905  0.091  0.627  0.145    0.065   

       관대함  부패에 대한인식    년도  
0    0.153     0.393  2019  
1    0.252     0.410  2019  
2    0.271     0.341  2019  
3    0.354     0.118  2019  
4    0.322     0.298  2019  
..     ...       ...   ...  
307  0.083     0.064  2018  
308  0.270     0.097  2018  
309  0.224     0.106  2018  
310  0.218     0.038  2018  
311  0.149     0.076  2018  

[312 rows x 10 columns]

##### 41. 데이터는 2018년도와 2019년도의 전세계 행복 지수를 표현한다. 각년도의 행복랭킹 10위를 차지한 나라의 행복점수의 평균을 구하여라

In [58]:
df[df["행복랭킹"] == 10]["점수"].mean()

np.float64(7.259)

##### 42. 데이터는 2018년도와 2019년도의 전세계 행복 지수를 표현한다. 각년도의 행복랭킹 50위이내의 나라들의 각각의 행복점수 평균을 데이터프레임으로 표시하라

In [59]:
df[df["행복랭킹"] <= 50][["년도", "점수"]].groupby("년도").mean()

점수
년도           
2018  6.64678
2019  6.67002

##### 43. 2018년도 데이터들만 추출하여 행복점수와 부패에 대한 인식에 대한 상관계수를 구하여라

In [60]:
df[df["년도"] == 2018][["점수", "부패에 대한인식"]].corr().iloc[0,1]

np.float64(0.40529152271510027)

##### 44. 2018년도와 2019년도의 행복랭킹이 변화하지 않은 나라명의 수를 구하여라

In [61]:
len(df[["행복랭킹", "나라명"]]) - len(df[["행복랭킹", "나라명"]].drop_duplicates())

15

##### 45. ❌ 2019년도 데이터들만 추출하여 각변수간 상관계수를 구하고 내림차순으로 정렬한 후 상위 5개를 데이터 프레임으로 출력하라. 컬럼명은 v1,v2,corr으로 표시하라

In [62]:
df[["행복랭킹", "점수", "상대GDP", "사회적지원", "행복기대치", "선택의 자유도", "관대함", "부패에 대한인식"]].corr().unstack()

행복랭킹      행복랭킹        1.000000
          점수         -0.990325
          상대GDP      -0.803757
          사회적지원      -0.752372
          행복기대치      -0.757395
                        ...   
부패에 대한인식  사회적지원       0.200387
          행복기대치       0.293732
          선택의 자유도     0.443004
          관대함         0.344607
          부패에 대한인식    1.000000
Length: 64, dtype: float64

In [63]:
target = df[df["년도"] == 2019]
res = target[["행복랭킹", "점수", "상대GDP", "사회적지원", "행복기대치", "선택의 자유도", "관대함", "부패에 대한인식"]].corr().unstack().reset_index().dropna()
res = res[res[0] != 1].sort_values(0, ascending = False).drop_duplicates(0)
ans = res.head(5).reset_index(drop = True)
ans.columns = ["v1", "v2", "corr"]
ans

v1     v2      corr
0  상대GDP  행복기대치  0.835462
1  상대GDP     점수  0.793883
2  행복기대치     점수  0.779883
3     점수  사회적지원  0.777058
4  사회적지원  상대GDP  0.754906

##### 46. ❌ 각 년도별 하위 행복점수의 하위 5개 국가의 평균 행복점수를 구하여라

In [64]:
df.groupby("년도").tail(5).groupby("년도", as_index = False)["점수"].mean()

년도      점수
0  2018  3.1800
1  2019  3.1408

##### 47. 2019년 데이터를 추출하고 해당데이터의 상대 GDP 평균 이상의 나라들과 평균 이하의 나라들의 행복점수 평균을 각각 구하고 그 차이값을 출력하라

In [65]:
target = df[df["년도"] == 2019]
over = target[df["상대GDP"] >= target["상대GDP"].mean()]["점수"].mean()
under = target[df["상대GDP"] <= target["상대GDP"].mean()]["점수"].mean()

print(over - under)

1.390483957219251


C:\Users\cho03\AppData\Local\Temp\ipykernel_31996\580180198.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  over = target[df["상대GDP"] >= target["상대GDP"].mean()]["점수"].mean()
C:\Users\cho03\AppData\Local\Temp\ipykernel_31996\580180198.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  under = target[df["상대GDP"] <= target["상대GDP"].mean()]["점수"].mean()


##### 48. ❌ 각 년도의 부패에 대한 인식을 내림차순 정렬했을때 상위 20개 국가의 부패에 대한 인식의 평균을 구하여라

In [66]:
df.sort_values(["년도", "부패에 대한인식"], ascending = False).groupby("년도").head(20).groupby("년도")["부패에 대한인식"].mean()

년도
2018    0.3267
2019    0.3201
Name: 부패에 대한인식, dtype: float64

##### 49. 2018년도 행복랭킹 50위 이내에 포함됐다가 2019년 50위 밖으로 밀려난 국가의 숫자를 구하여라

In [67]:
len(set(df[(df["년도"] == 2018) & (df["행복랭킹"] <= 50)]["나라명"])  -set(df[(df["년도"] == 2019) & (df["행복랭킹"] <= 50)]["나라명"]))

4

##### 50. ❌ 2018년,2019년 모두 기록이 있는 나라들 중 년도별 행복점수가 가장 증가한 나라와 그 증가 수치는?

In [68]:
count = df["나라명"].value_counts()
target = count[count >= 2].index # 2018, 2019 모두 기록이 있는 나라명

df2 = df.copy()
multiple = df2[df2["나라명"].isin(target)].reset_index(drop = True) # 2018, 2019 모두 기록이 있는 나라 데이터
multiple.loc[multiple["년도"] == 2018, "점수"] = multiple[multiple["년도"] == 2018]["점수"].values * (-1)
multiple.groupby("나라명").sum()["점수"].sort_values().to_frame().iloc[-1]

점수    0.87
Name: Burundi, dtype: float64

# 지역구 에너지 소비량 데이터2
데이터 출처 : https://archive.ics.uci.edu/ml/datasets/Power+consumption+of+Tetouan+city (참고, 데이터 수정)  
데이터 설명 : 기온, 습도,바람풍속에 따른 도시의 3개 지역구의 에너지 소비량  
data url = https://raw.githubusercontent.com/Datamanim/datarepo/main/consum/Tetuan City power consumption.csv

In [69]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/consum/Tetuan%20City%20power%20consumption.csv")
df.head()

DateTime  Temperature  Humidity  Wind Speed  general diffuse flows  \
0  1/1/2017 0:00        6.559      73.8       0.083                  0.051   
1  1/1/2017 0:10        6.414      74.5       0.083                  0.070   
2  1/1/2017 0:20        6.313      74.5       0.080                  0.062   
3  1/1/2017 0:30        6.121      75.0       0.083                  0.091   
4  1/1/2017 0:40        5.921      75.7       0.081                  0.048   

   diffuse flows  Zone 1 Power Consumption  Zone 2  Power Consumption  \
0          0.119               34055.69620                16128.87538   
1          0.085               29814.68354                19375.07599   
2          0.100               29128.10127                19006.68693   
3          0.096               28228.86076                18361.09422   
4          0.085               27335.69620                17872.34043   

   Zone 3  Power Consumption  
0                20240.96386  
1                20131.08434  
2                19668.43373  
3                18899.27711  
4                18442.40964

##### 51. DateTime컬럼을 통해 각 월별로 몇개의 데이터가 있는지 데이터 프레임으로 구하여라

In [70]:
df["DateTime"] = pd.to_datetime(df["DateTime"])
df['DateTime'].dt.month.value_counts().sort_index().to_frame()

count
DateTime       
1          4464
2          4032
3          4464
4          4320
5          4464
6          4320
7          4464
8          4464
9          4320
10         4464
11         4320
12         4320

##### 52. 3월달의 각 시간대별 온도의 평균들 중 가장 낮은 시간대의 온도를 출력하라

In [71]:
target = df[df["DateTime"].dt.month == 3]
target.groupby(target["DateTime"].dt.hour)["Temperature"].mean().sort_values().values[0]

np.float64(11.506612903225806)

##### 53. 3월달의 각 시간대별 온도의 평균들 중 가장 높은 시간대의 온도를 출력하라

In [72]:
target.groupby(target["DateTime"].dt.hour)["Temperature"].mean().sort_values().values[-1]

np.float64(18.393602150537635)

##### 54. Zone 1 Power Consumption 컬럼의 value값의 크기가 Zone 2 Power Consumption 컬럼의 value값의 크기보다 큰 데이터들의 Humidity의 평균을 구하여라

In [73]:
df[df["Zone 1 Power Consumption"] > df["Zone 2  Power Consumption"]]["Humidity"].mean()

np.float64(68.23624448055051)

##### 55. 각 zone의 에너지 소비량의 상관관계를 구해서 데이터 프레임으로 표기하라

In [74]:
df[["Zone 1 Power Consumption", "Zone 2  Power Consumption", "Zone 3  Power Consumption"]].corr()

Zone 1 Power Consumption  \
Zone 1 Power Consumption                   1.000000   
Zone 2  Power Consumption                  0.834519   
Zone 3  Power Consumption                  0.750733   

                           Zone 2  Power Consumption  \
Zone 1 Power Consumption                    0.834519   
Zone 2  Power Consumption                   1.000000   
Zone 3  Power Consumption                   0.570932   

                           Zone 3  Power Consumption  
Zone 1 Power Consumption                    0.750733  
Zone 2  Power Consumption                   0.570932  
Zone 3  Power Consumption                   1.000000

##### 56. Temperature의 값이 10미만의 경우 A, 10이상 20미만의 경우 B,20이상 30미만의 경우 C, 그 외의 경우 D라고 할때 각 단계의 데이터 숫자를 구하여라

In [75]:
def split_data(x):
    if x < 10: return "A"
    elif x < 20: return "B"
    elif x < 30: return "C"
    else: return "D"
    
df["sp"] = df["Temperature"].map(split_data)
df["sp"].value_counts()

sp
B    26993
C    21105
A     2874
D     1444
Name: count, dtype: int64

##### 57. 6월 데이터중 12시의 Temperature의 표준편차를 구하여라

In [76]:
df[(df["DateTime"].dt.month == 6) & (df["DateTime"].dt.hour == 12)]["Temperature"].std()

np.float64(2.0499417827951025)

##### 58. 6월 데이터중 12시의 Temperature의 분산을 구하여라

In [77]:
df[(df["DateTime"].dt.month == 6) & (df["DateTime"].dt.hour == 12)]["Temperature"].var()

np.float64(4.202261312849163)

##### 59. Temperature의 평균이상의 Temperature의 값을 가지는 데이터를 Temperature를 기준으로 정렬 했을때 4번째 행의 Humidity 값은?

In [78]:
df[df["Temperature"] >= df["Temperature"].mean()].sort_values("Temperature")["Humidity"].values[3]

np.float64(79.7)

##### 60. Temperature의 중간값 이상의 Temperature의 값을 가지는 데이터를Temperature를 기준으로 정렬 했을때 4번째 행의 Humidity 값은?

In [79]:
df[df["Temperature"] >= df["Temperature"].median()].sort_values("Temperature")["Humidity"].values[3]

np.float64(58.98)

# 포켓몬 정보 데이터
데이터 출처 : https://www.kaggle.com/abcsds/pokemon (참고, 데이터 수정)  
데이터 설명 : 포켓몬 정보  
data url = https://raw.githubusercontent.com/Datamanim/datarepo/main/pok/Pokemon.csv

In [80]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/pok/Pokemon.csv")
df

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

##### 61. Legendary 컬럼은 전설포켓몬 유무를 나타낸다.전설포켓몬과 그렇지 않은 포켓몬들의 HP평균의 차이를 구하여라

In [81]:
leg = df[df["Legendary"] == True]["HP"].mean()
nleg = df[df["Legendary"] == False]["HP"].mean()

leg - nleg

np.float64(25.55614861329147)

##### 62. Type 1은 주속성 Type 2 는 부속성을 나타낸다. 가장 많은 부속성 종류는 무엇인가?

In [82]:
df.value_counts("Type 2").index[0]

'Flying'

In [83]:
df.groupby("Type 2").size().sort_values().index[-1]

'Flying'

##### 63. 가장 많은 Type 1 의 종의 평균 Attack 을 평균 Defense로 나눈값은?

In [84]:
max = df['Type 1'].value_counts().index[0]

df[df["Type 1"] == max]["Attack"].mean() / df[df["Type 1"]== max]["Defense"].mean()

np.float64(1.0165238678090576)

##### 64. 포켓몬 세대(Generation) 중 가장많은 Legendary를 보유한 세대는 몇세대인가?

In [85]:
print(df[df["Legendary"] == True]["Generation"].value_counts().index[0])

3


##### 65. ‘HP’, ‘Attack’, ‘Defense’, ‘Sp. Atk’, ‘Sp. Def’, ‘Speed’ 간의 상관 계수중 가장 절댓값이 큰 두 변수와 그 값을 구하여라

In [86]:
target = df[["HP", "Attack", "Defense", "Sp. Atk", "Sp. Def", "Speed"]].corr().unstack().reset_index()
target[target[0] != 1].sort_values(0, ascending = False).iloc[0]

level_0     Defense
level_1     Sp. Def
0          0.510747
Name: 16, dtype: object

##### 66. 각 Generation의 Attack으로 오름차순 정렬시 상위 3개 데이터들(18개)의 Attack의 전체 평균을 구하여라

In [87]:
df.sort_values(["Generation", "Attack"]).groupby("Generation").head(3)["Attack"].mean()

np.float64(19.5)

##### 67. 각 Generation의 Attack으로 내림차순 정렬시 상위 5개 데이터들(30개)의 Attack의 전체 평균을 구하여라

In [88]:
df.sort_values(["Generation", "Attack"], ascending = False).groupby("Generation").head(5)["Attack"].mean()

np.float64(157.23333333333332)

##### 68. 가장 흔하게 발견되는 (Type1 , Type2) 의 쌍은 무엇인가?

In [89]:
df[['Type 1','Type 2']].value_counts().head(1)

Type 1  Type 2
Normal  Flying    24
Name: count, dtype: int64

##### 69. 한번씩만 존재하는 (Type1 , Type2)의 쌍의 갯수는 몇개인가?

In [90]:
target = df[['Type 1','Type 2']].value_counts()
len(target[target == 1])

39

##### 70. ❌ 한번씩만 존재하는 (Type1 , Type2)의 쌍을 각 세대(Generation)은 각각 몇개씩 가지고 있는가?

In [91]:
target2 = target[target == 1]

lst = []
for value in target2.reset_index().values:
    t1 = value[0]
    t2 = value[1]
    
    sp = df[(df["Type 1"] == t1) & (df["Type 2"] == t2)]
    lst.append(sp)

result = pd.concat(lst).reset_index(drop = True)["Generation"].value_counts().sort_index()
print(result)

Generation
1     1
2     4
3     5
4    13
5     7
6     9
Name: count, dtype: int64


# 대한민국 체력장 데이터
데이터 출처 : 국민체육진흥공단 (문화 빅데이터플랫폼) (참고, 데이터 수정)  
데이터 설명 : 대한민국 국민 체력장 평가  
data url = https://raw.githubusercontent.com/Datamanim/datarepo/main/body/body.csv

In [92]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/body/body.csv")

##### 71. 전체데이터의 수축기혈압(최고) - 이완기혈압(최저)의 평균을 구하여라

In [93]:
(df["수축기혈압(최고) : mmHg"] - df["이완기혈압(최저) : mmHg"]).mean()

np.float64(52.19539414750672)

##### 72. 50~59세의 신장평균을 구하여라

In [94]:
df[(df["측정나이"] >= 50) & (df["측정나이"] <= 59)]["신장 : cm"].mean()

np.float64(164.07490107405314)

##### 73. 연령대 (20~29 : 20대 …) 별 인원수를 구하여라

In [95]:
df["연령대"]  = df["측정나이"] // 10 * 10
df["연령대"].value_counts()

연령대
20.0    5831
30.0    2660
40.0    1801
50.0    1769
60.0    1335
Name: count, dtype: int64

##### 74. 연령대 (20~29 : 20대 …) 별 등급의 숫자를 데이터 프레임으로 표현하라

In [96]:
df.groupby(['연령대','등급'], as_index = False).size()

연령대 등급  size
0   20.0  A  1585
1   20.0  B  1443
2   20.0  C  1455
3   20.0  D  1348
4   30.0  A   743
5   30.0  B   697
6   30.0  C   626
7   30.0  D   594
8   40.0  A   386
9   40.0  B   428
10  40.0  C   455
11  40.0  D   532
12  50.0  A   321
13  50.0  B   410
14  50.0  C   474
15  50.0  D   564
16  60.0  A   314
17  60.0  B   371
18  60.0  C   339
19  60.0  D   311

In [97]:
df[["연령대", "등급"]].value_counts().reset_index(name = "size")

연령대 등급  size
0   20.0  A  1585
1   20.0  C  1455
2   20.0  B  1443
3   20.0  D  1348
4   30.0  A   743
5   30.0  B   697
6   30.0  C   626
7   30.0  D   594
8   50.0  D   564
9   40.0  D   532
10  50.0  C   474
11  40.0  C   455
12  40.0  B   428
13  50.0  B   410
14  40.0  A   386
15  60.0  B   371
16  60.0  C   339
17  50.0  A   321
18  60.0  A   314
19  60.0  D   311

##### 75. 남성 중 A등급과 D등급의 체지방률 평균의 차이(큰 값에서 작은 값의 차)를 구하여라

In [98]:
target = df[df["측정회원성별"] == "M"]
abs(target[target["등급"] == "A"]["체지방율 : %"].mean() - target[target["등급"] == "D"]["체지방율 : %"].mean())

np.float64(7.932086486137521)

##### 76. 여성 중 A등급과 D등급의 체중의 평균의 차이(큰 값에서 작은 값의 차)를 구하여라

In [99]:
target = df[df["측정회원성별"] == "F"]
abs(target[target["등급"] == "A"]["체중 : kg"].mean() - target[target["등급"] == "D"]["체중 : kg"].mean())

np.float64(5.176211590296504)

##### 77. bmi는 자신의 몸무게(kg)를 키의 제곱(m)으로 나눈값이다. 데이터의 bmi 를 구한 새로운 컬럼을 만들고 남성의 bmi 평균을 구하여라

In [100]:
df["bmi"] = df["체중 : kg"] / ((df["신장 : cm"] / 100) ** 2)
df[df["측정회원성별"] == "M"]["bmi"].mean()

np.float64(24.46134409819308)

##### 78. bmi보다 체지방율이 높은 사람들의 체중평균을 구하여라

In [101]:
df[df["체지방율 : %"] > df["bmi"]]["체중 : kg"].mean()

np.float64(61.7408806392542)

##### 79. 남성과 여성의 악력 평균의 차이를 구하여라

In [102]:
df[df["측정회원성별"] == "M"]["악력D : kg"].mean() - df[df["측정회원성별"] == "F"]["악력D : kg"].mean()

np.float64(17.55954185047465)

##### 80. 남성과 여성의 교차윗몸일으키기 횟수의 평균의 차이를 구하여라

In [103]:
df[df["측정회원성별"] == "M"]["교차윗몸일으키기 : 회"].mean() - df[df["측정회원성별"] == "F"]["교차윗몸일으키기 : 회"].mean()

np.float64(14.243156833157627)

# 기온 강수량 데이터
데이터 출처 : https://data.kma.go.kr/cmmn/static/staticPage.do?page=intro  
데이터 설명 : 2020년도 이화동(서울) , 수영동(부산)의 시간단위의 기온과 강수량  
데이터 url : https://raw.githubusercontent.com/Datamanim/datarepo/main/weather/weather2.csv

In [104]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/weather/weather2.csv")
df

time  이화동강수  이화동기온  수영동강수  수영동기온
0     2020-01-01 00:00:00    0.1   -3.9    0.0    2.5
1     2020-01-01 01:00:00    0.0   -3.1    0.0    3.4
2     2020-01-01 02:00:00    0.0   -1.8    0.0    3.8
3     2020-01-01 03:00:00    0.0   -0.8    0.0    4.6
4     2020-01-01 04:00:00    0.0   -0.1    0.0    5.1
...                   ...    ...    ...    ...    ...
8779  2020-12-31 19:00:00    0.0   -9.5    0.0   -4.5
8780  2020-12-31 20:00:00    0.0   -9.5    0.0   -5.0
8781  2020-12-31 21:00:00    0.0   -9.3    0.0   -6.5
8782  2020-12-31 22:00:00    0.0   -9.2    0.0   -4.6
8783  2020-12-31 23:00:00    0.0   -7.8    0.0   -0.2

[8784 rows x 5 columns]

##### 81. 여름철(6월,7월,8월) 이화동이 수영동보다 높은 기온을 가진 시간대는 몇개인가?

In [105]:
df["time"] = pd.to_datetime(df["time"])
target = df[df["time"].dt.month.isin([6, 7, 8])]
len(target[target["이화동기온"] > target["수영동기온"]])

1415

##### 82. 이화동과 수영동의 최대강수량의 시간대를 각각 구하여라

In [106]:
print(df[df["이화동강수"] == df["이화동강수"].max()]["time"].values, df[df["수영동강수"] == df["수영동강수"].max()]["time"].values)

['2020-09-30T09:00:00.000000000'] ['2020-07-23T12:00:00.000000000']


# 서비스 이탈예측 데이터
데이터 출처 : https://www.kaggle.com/shubh0799/churn-modelling 에서 변형  
데이터 설명 : 고객의 신상정보 데이터를 통한 회사 서비스 이탈 예측 (종속변수 : Exited)  
Data url : https://raw.githubusercontent.com/Datamanim/datarepo/main/churn/train.csv

In [107]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/churn/train.csv")
df

RowNumber  CustomerId     Surname  CreditScore Geography  Gender  Age  \
0          6842    15793491  Cherkasova          714   Germany    Male   26   
1          8963    15607874       Keane          687    France    Male   38   
2          7047    15737627      Rivero          589   Germany  Female   20   
3          7503    15697844  Whitehouse          721     Spain  Female   32   
4          3439    15722404   Carpenter          445    France  Female   30   
...         ...         ...         ...          ...       ...     ...  ...   
7994       1607    15714080      Goliwe          566   Germany  Female   40   
7995       7744    15741643      Chiang          777   Germany    Male   35   
7996       4231    15566660        Cole          670    France  Female   41   
7997       8870    15733597        Y?an          669    France  Female   41   
7998        523    15595766       Watts          527     Spain    Male   37   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          3  119545.48              2          1               0   
1          0  144450.58              1          0               1   
2          2  121093.29              2          1               0   
3         10       0.00              1          1               0   
4          3       0.00              2          1               1   
...      ...        ...            ...        ...             ...   
7994       2   97001.36              2          1               0   
7995       7  122917.69              1          1               1   
7996      10       0.00              3          1               0   
7997       0  150219.41              2          0               0   
7998       5   93722.73              2          1               1   

      EstimatedSalary  Exited  
0            65482.94       0  
1           137276.83       0  
2             3529.72       0  
3           136119.96       1  
4           127939.19       0  
...               ...     ...  
7994        154486.01       0  
7995         76169.68       0  
7996         81602.02       0  
7997        107839.03       0  
7998        139093.73       0  

[7999 rows x 14 columns]

##### 83. ❌ 남성 이탈(Exited)이 가장 많은 국가(Geography)는 어디이고 이탈 인원은 몇명인가?

In [108]:
df[df["Gender"] == "Male"].groupby("Geography").sum()["Exited"].sort_values(ascending = False).head(1)

Geography
Germany    287
Name: Exited, dtype: int64

##### 84. 카드를 소유(HasCrCard ==1)하고 있으면서 활성멤버(IsActiveMember ==1) 인 고객들의 평균 나이를 소숫점이하 4자리까지 구하여라

In [109]:
round(df.loc[(df["HasCrCard"] == 1) & (df["IsActiveMember"] == 1)]["Age"].mean(), 4)

np.float64(39.6102)

##### 85. Balance 값이 중간값 이상을 가지는 고객들의 CreditScore의 표준편차를 소숫점이하 3자리까지 구하여라

In [110]:
round(df[df["Balance"] >= df["Balance"].median()]["CreditScore"].std(), 3)

np.float64(97.295)

# 성인 건강검진 데이터
데이터 출처 :https://www.data.go.kr/data/15007122/fileData.do (참고, 데이터 수정)  
데이터 설명 : 2018년도 성인의 건강검 진데이터 (흡연상태 1- 흡연, 0-비흡연 )  
데이터 url : https://raw.githubusercontent.com/Datamanim/datarepo/main/smoke/train.csv

In [111]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/smoke/train.csv")
df

성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg단위)  허리둘레  시력(좌)  시력(우)  청력(좌)  \
0        F           55        145         55  73.0    0.7    0.5    1.0   
1        M           40        180         55  74.0    1.5    1.5    1.0   
2        F           55        150         50  72.0    1.0    0.2    1.0   
3        M           40        170         85  89.0    1.0    0.8    1.0   
4        F           40        155         45  62.0    0.5    1.0    1.0   
...    ...          ...        ...        ...   ...    ...    ...    ...   
44548    F           60        145         50  87.0    1.0    1.0    1.0   
44549    M           40        175         75  88.0    1.5    1.2    1.0   
44550    F           60        150         55  82.0    0.9    0.9    1.0   
44551    M           40        170         50  69.0    0.4    1.2    1.0   
44552    F           40        155         50  72.5    1.0    1.0    1.0   

       청력(우)  수축기혈압  ...   혈색소  요단백  혈청크레아티닌  (혈청지오티)AST  (혈청지오티)ALT  감마지티피  \
0        1.0  129.0  ...  12.9  1.0      0.9       172.0       209.0   15.0   
1        1.0  102.0  ...  15.6  1.0      0.8        30.0        19.0   23.0   
2        1.0  116.0  ...  14.1  1.0      1.0        31.0        19.0   10.0   
3        1.0  124.0  ...  16.7  1.0      1.0        26.0        38.0  108.0   
4        1.0  127.0  ...  12.9  1.0      0.7        20.0        11.0   13.0   
...      ...    ...  ...   ...  ...      ...         ...         ...    ...   
44548    1.0  140.0  ...  12.3  1.0      0.5        24.0        32.0   15.0   
44549    1.0  139.0  ...  15.3  1.0      0.8        62.0       106.0   49.0   
44550    1.0  124.0  ...  14.1  1.0      0.5        38.0        34.0   26.0   
44551    1.0  117.0  ...  14.2  1.0      0.9        22.0        17.0   55.0   
44552    1.0  109.0  ...  11.7  1.0      0.6        16.0        10.0    7.0   

       흡연상태  구강검진수검여부  치아우식증유무  치석  
0         0         Y      0.0   Y  
1         1         Y      0.0   Y  
2         0         Y      0.0   N  
3         1         Y      1.0   Y  
4         0         Y      0.0   N  
...     ...       ...      ...  ..  
44548     0         Y      1.0   Y  
44549     1         Y      0.0   Y  
44550     0         Y      0.0   Y  
44551     1         Y      0.0   N  
44552     0         Y      0.0   N  

[44553 rows x 26 columns]

##### 86. 수축기혈압과 이완기 혈압기 수치의 차이를 새로운 컬럼(‘혈압차’) 으로 생성하고, 연령대 코드별 각 그룹 중 ‘혈압차’ 의 분산이 5번째로 큰 연령대 코드를 구하여라

In [112]:
df["혈압차"] = df["수축기혈압"] - df["이완기혈압"]
df.groupby("연령대코드(5세단위)")["혈압차"].var().sort_values(ascending = False).index[4]

np.int64(60)

##### 87. ❌ 비만도를 나타내는 지표인 WHtR는 허리둘레 / 키로 표현한다. 일반적으로 0.58이상이면 비만으로 분류한다. 데이터중 WHtR 지표상 비만인 인원의 남/여 비율을 구하여라

In [113]:
df["WHtR"] = df["허리둘레"] / df["신장(5Cm단위)"]
res = df[df["WHtR"] >= 0.58]["성별코드"].value_counts()
res["M"] / res["F"]

np.float64(1.1693877551020408)

# 자동차 보험가입 예측데이터
데이터 출처 :https://www.kaggle.com/anmolkumar/health-insurance-cross-sell-prediction(참고, 데이터 수정)  
데이터 설명 : 자동차 보험 가입 예측  
Data url : https://raw.githubusercontent.com/Datamanim/datarepo/main/insurance/train.csv

In [114]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/insurance/train.csv")
df

id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0       188957  Female   25                1         17.0                   1   
1       275631    Male   20                1         22.0                   0   
2       329036  Female   40                1         41.0                   0   
3       227288  Female   28                1         30.0                   1   
4       300441  Female   50                1         28.0                   0   
...        ...     ...  ...              ...          ...                 ...   
304882  351529    Male   75                1         33.0                   0   
304883  287917    Male   57                1         28.0                   1   
304884   21023  Female   40                1         35.0                   0   
304885   67340    Male   24                1         41.0                   0   
304886  127164  Female   22                1         50.0                   1   

       Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  \
0         < 1 Year             No         38785.0                 152.0   
1         < 1 Year             No          2630.0                 160.0   
2         1-2 Year            Yes         33165.0                 124.0   
3         < 1 Year             No         29038.0                 152.0   
4         1-2 Year            Yes         31325.0                 124.0   
...            ...            ...             ...                   ...   
304882    1-2 Year            Yes         32412.0                  26.0   
304883    1-2 Year             No         44602.0                  52.0   
304884    1-2 Year            Yes         38779.0                  26.0   
304885    < 1 Year            Yes         26669.0                 152.0   
304886    < 1 Year             No         35063.0                 152.0   

        Vintage  Response  
0            59         0  
1           113         0  
2           188         0  
3           233         0  
4           258         0  
...         ...       ...  
304882      296         1  
304883      135         0  
304884      293         0  
304885       29         0  
304886      184         0  

[304887 rows x 12 columns]

##### 88. Vehicle_Age 값이 2년 이상인 사람들만 필터링 하고 그중에서 Annual_Premium 값이 전체 데이터의 중간값 이상인 사람들을 찾고, 그들의 Vintage값의 평균을 구하여라

In [115]:
df[(df["Vehicle_Age"]=="> 2 Years") & (df["Annual_Premium"] >= df["Annual_Premium"].median())]["Vintage"].mean()

np.float64(154.43647182359118)

##### 89. ❌ vehicle_age에 따른 각 성별(gender)그룹의 Annual_Premium값의 평균을 구하여 아래 테이블과 동일하게 구현하라

In [116]:
res = df.groupby(["Gender", "Vehicle_Age"], as_index = False)["Annual_Premium"].mean()
res.pivot_table(index = "Vehicle_Age", columns = "Gender", values = "Annual_Premium")

Gender             Female          Male
Vehicle_Age                            
1-2 Year     30762.245001  30413.088469
< 1 Year     29972.286702  30310.982212
> 2 Years    36108.366374  35303.870627

# 핸드폰 가격 예측데이터
데이터 출처 :https://www.kaggle.com/iabhishekofficial/mobile-price-classification?select=train.csv(참고, 데이터 수정)  
데이터 설명 : 핸드폰 가격예측 (price_range컬럼 0(저렴) ~3(매우비쌈) 범위 )  
Data url https://raw.githubusercontent.com/Datamanim/datarepo/main/mobile/train.csv

In [117]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/mobile/train.csv")
df

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0               842     0          2.2         0   1       0           7   
1              1021     1          0.5         1   0       1          53   
2               563     1          0.5         1   2       1          41   
3               615     1          2.5         0   0       0          10   
4              1821     1          1.2         0  13       1          44   
...             ...   ...          ...       ...  ..     ...         ...   
1995            794     1          0.5         1   0       1           2   
1996           1965     1          2.6         1   0       0          39   
1997           1911     0          0.9         1   1       1          36   
1998           1512     0          0.9         0   4       1          46   
1999            510     1          2.0         1   5       1          45   

      m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  \
0       0.6        188        2  ...         20       756  2549     9     7   
1       0.7        136        3  ...        905      1988  2631    17     3   
2       0.9        145        5  ...       1263      1716  2603    11     2   
3       0.8        131        6  ...       1216      1786  2769    16     8   
4       0.6        141        2  ...       1208      1212  1411     8     2   
...     ...        ...      ...  ...        ...       ...   ...   ...   ...   
1995    0.8        106        6  ...       1222      1890   668    13     4   
1996    0.2        187        4  ...        915      1965  2032    11    10   
1997    0.7        108        8  ...        868      1632  3057     9     1   
1998    0.1        145        5  ...        336       670   869    18    10   
1999    0.9        168        6  ...        483       754  3919    19     4   

      talk_time  three_g  touch_screen  wifi  price_range  
0            19        0             0     1            1  
1             7        1             1     0            2  
2             9        1             1     0            2  
3            11        1             0     0            2  
4            15        1             1     0            1  
...         ...      ...           ...   ...          ...  
1995         19        1             1     0            0  
1996         16        1             1     1            2  
1997          5        1             1     0            3  
1998         19        1             1     1            0  
1999          2        1             1     1            3  

[2000 rows x 21 columns]

##### 90. price_range 의 각 value를 그룹핑하여 각 그룹의 n_cores 의 빈도가 가장높은 value와 그 빈도수를 구하여라

##### 91. price_range 값이 3인 그룹에서 상관관계가 2번째로 높은 두 컬럼과 그 상관계수를 구하여라

# 비행탑승 경험 만족도 데이터
데이터 출처 :https://www.kaggle.com/teejmahal20/airline-passenger-satisfaction?select=train.csv (참고, 데이터 수정)  
데이터 설명 : 비행탑승 경험 만족도 (satisfaction 컬럼 : ‘neutral or dissatisfied’ or satisfied ) (83123, 24) shape  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/airline/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/airline/train.csv")
df

##### 92. Arrival Delay in Minutes 컬럼이 결측치인 데이터들 중 ‘neutral or dissatisfied’ 보다 ‘satisfied’의 수가 더 높은 Class는 어디 인가?

# 수질 음용성 여부 데이터
데이터 출처 :https://www.kaggle.com/adityakadiwal/water-potability  
데이터 설명 : 수질 음용성 여부 (Potablillity 컬럼 : 0 ,1 )  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/waters/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/waters/train.csv")
df

##### 93. ph값은 상당히 많은 결측치를 포함한다. 결측치를 제외한 나머지 데이터들 중 사분위값 기준 하위 25%의 값들의 평균값은?

# 의료 비용 예측 데이터
데이터 출처 :https://www.kaggle.com/mirichoi0218/insurance/code(참고, 데이터 수정)  
데이터 설명 : 의료비용 예측문제  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/MedicalCost/train.csv

In [ ]:
train = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/MedicalCost/train.csv")
train

##### 94. 흡연자와 비흡연자 각각 charges의 상위 10% 그룹의 평균의 차이는?

# 킹카운티 주거지 가격예측문제 데이터
데이터 출처 :https://www.kaggle.com/harlfoxem/housesalesprediction (참고, 데이터 수정)  
데이터 설명 : 킹카운티 주거지 가격 예측문제  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/kingcountyprice/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/kingcountyprice//train.csv")
df

##### 95. bedrooms 의 빈도가 가장 높은 값을 가지는 데이터들의 price의 상위 10%와 하위 10%값의 차이를 구하여라

# 대학원 입학가능성 데이터
데이터 출처 :https://www.kaggle.com/mohansacharya/graduate-admissions(참고, 데이터 수정)  
데이터 설명 : 대학원 입학 가능성 예측  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/admission/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/admission/train.csv")
df

##### 96. Serial No. 컬럼을 제외하고 ‘Chance of Admit’을 종속변수, 나머지 변수를 독립변수라 할때, 랜덤포레스트를 통해 회귀 예측을 할 떄 변수중요도 값을 출력하라 (시드값에 따라 순서는 달라질수 있음)

# 레드 와인 퀄리티 예측 데이터
데이터 출처 :https://www.kaggle.com/uciml/red-wine-quality-cortez-et-al-2009(참고, 데이터 수정)  
데이터 설명 : 레드 와인 퀄리티 예측문제  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/redwine/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/redwine/train.csv")
df

##### 97. quality 값이 3인 그룹과 8인 데이터그룹의 각 컬럼별 독립변수의 표준편차 값의 차이를 구할때 그값이 가장 큰 컬럼명을 구하여라

# 약물 분류 데이터
데이터 출처 :https://www.kaggle.com/prathamtripathi/drug-classification(참고, 데이터 수정)  
데이터 설명 : 투약하는 약을 분류  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/drug/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/drug/train.csv")
df

##### 98. 남성들의 연령대별 (10살씩 구분 0~9세 10~19세 …) Na_to_K값의 평균값을 구해서 데이터 프레임으로 표현하여라

# 사기회사 분류 데이터
데이터 출처 :https://www.kaggle.com/sid321axn/audit-data(참고, 데이터 수정)  
데이터 설명 : 사기회사 분류  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/audit/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/audit/train.csv")
df

##### 99. 데이터의 Risk 값에 따른 score_a와 score_b의 평균값을 구하여라

# 센서데이터 동작유형 분류 데이터
데이터 출처 :https://www.kaggle.com/kyr7plus/emg-4(참고, 데이터 수정)  
데이터 설명 : 센서데이터로 동작 유형 분류  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/muscle/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/muscle/train.csv")
df

##### 100. pose값에 따른 각 motion컬럼의 중간값의 가장 큰 차이를 보이는 motion컬럼은 어디이며 그값은?

# 현대 차량 가격 분류문제 데이터
데이터 출처 :https://www.kaggle.com/mysarahmadbhat/hyundai-used-car-listing(참고, 데이터 수정)  
데이터 설명 : 현대 차량가격 분류문제  
Data url :hhttps://raw.githubusercontent.com/Datamanim/datarepo/main/hyundai/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/hyundai/train.csv")
df

##### 101. 정보(row수)가 가장 많은 상위 3차종의 price값의 각 평균값은?

# 당뇨여부판단 데이터
데이터 출처 :https://www.kaggle.com/pritsheta/diabetes-dataset(참고, 데이터 수정)  
데이터 설명 : 당뇨여부 판단하기  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/diabetes/train.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/diabetes/train.csv")
df

##### 102. Outcome 값에 따른 각 그룹의 각 컬럼의 평균 차이를 구하여라

# 넷플릭스 주식 데이터
데이터 출처 :https://www.kaggle.com/pritsheta/netflix-stock-data-from-2002-to-2021(참고, 데이터 수정)  
데이터 설명 : 넷플릭스 주식데이터  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/nflx/NFLX.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nflx/NFLX.csv")
df

##### 103. 매년 5월달의 open가격의 평균값을 데이터 프레임으로 표현하라

# nba 선수 능력치 데이터
데이터 출처 :https://www.kaggle.com/datasets/vivovinco/nba-player-stats(참고, 데이터 수정)  
Data url :https://raw.githubusercontent.com/Datamanim/datarepo/main/nba/nba.csv

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nba/nba.csv", encoding = "latin", sep = ";")
df

SyntaxError: unterminated string literal (detected at line 1) (90618726.py, line 1)

##### 104. Tm 컬럼은 각 팀의 이름을 의미한다. TOR팀의 평균나이를 소수 4째 자리까지 구하여라

##### 105. Pos 컬럼은 포지션을 의미한다. 전체 선수 중 최소나이대의 선수들을 필터하고 그들 중 가장 많은 포지션은 무엇인지 확인하라

##### 106. 선수들의 이름은 first_name+ 공백 + last_name으로 이루어져 있다. 가장 많은 first_name은 무엇이며 몇 회 발생하는지 확인하라

##### 107. PTS컬럼은 경기당 평균득점수 이다. 각포지션별로 경기당 평균득점수의 평균을 구하여라

##### 108. PTS컬럼은 경기당 평균득점수 이다. 각포지션별로 경기당 평균득점수의 평균을 구하여라

##### 109. G컬럼은 참여한 경기의 숫자이다. 각 팀별로 가장 높은 경기참여 수를 가진 선수들의 경기 참여 숫자의 평균을 구하여라

##### 110. Tm의 값이 MIA이며 Pos는 C또는 PF인 선수의 MP값의 평균은?

##### 111. 전체 데이터중 G의 평균값의 1.5배 이상인 데이터들만 추출했을때 3P값의 평균은?

##### 112. Age의 평균 이상인 그룹과 평균 미만인 그룹간의 G값의 평균의 차이는?

##### 113. 평균나이가 가장 젊은 팀은 어디인가?

##### 114. Pos그룹별 평균 MP값을 구하여라